File: sdk/python-sdk/verity_sdk/protocols/WriteCredentialDefinition.py
from verity_sdk.protocols.Protocol import Protocol


class WriteCredentialDefinition(Protocol):
    MSG_FAMILY = 'write-cred-def'
    MSG_FAMILY_VERSION = '0.6'

    # Messages
    WRITE_CRED_DEF = 'write'
    STATUS = 'status-report'

    def __init__(self, name: str, schema_id: str, tag: str = None, revocation_details: dict = None):
        super().__init__(self.MSG_FAMILY, self.MSG_FAMILY_VERSION)
        self.name = name
        self.schema_id = schema_id
        self.tag = tag
        self.revocation_details = revocation_details

    async def write_msg(self, _):
        msg = self._get_base_message(self.WRITE_CRED_DEF)
        msg['name'] = self.name
        msg['schemaId'] = self.schema_id
        msg['tag'] = self.tag
        msg['revocationDetails'] = self.revocation_details
        return msg

    async def write_msg_packed(self, context):
        return await self.get_message_bytes(context, await self.write_msg(context))

    async def write(self, context):
        await self.send_message(context, await self.write_msg_packed(context))

File: sdk/nodejs-sdk/src/protocols/v0_6/WriteCredentialDefinition.js
'use strict'
const utils = require('../../utils')
const Protocol = require('../Protocol')

/**
 * An interface for controlling a 0.6 WriteCredentialDefinition protocol.
 * @extends Protocol
 */
class WriteCredentialDefinition extends Protocol {
  /**
   * Constructor for the 0.6 WriteCredentialDefinition object. This constructor creates an object that is
   * ready to write a credential definitions the ledger
   *
   * @param name The name of the new credential definition
   * @param schemaId The id of the schema this credential definition will be based on
   * @param tag An optional tag for the credential definition
   * @param revocation the revocation object defining revocation support.
   * @param threadId the thread id of the already started protocol
   * @return 0.6 WriteCredentialDefinition object
   *
   * @property {String} msgFamily - 'write-cred-def'
   * @property {String} msgFamilyVersion - '0.6'
   * @property {String} msgQualifier - 'Community Qualifier'
   * @property {String} this.msgNames.WRITE - 'write'
   */
  constructor (name, schemaId, tag = null, revocation = null, threadId = null) {
    const msgFamily = 'write-cred-def'
    const msgFamilyVersion = '0.6'
    const msgQualifier = utils.constants.EVERNYM_MSG_QUALIFIER
    super(msgFamily, msgFamilyVersion, msgQualifier, threadId)
    this.name = name
    this.schemaId = schemaId
    this.tag = tag
    this.revocation = revocation

    this.msgNames.WRITE_CRED_DEF = 'write'
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #write
     */
  async writeMsg (context) {
    var msg = this._getBaseMessage(this.msgNames.WRITE_CRED_DEF)
    msg.name = this.name
    msg.schemaId = this.schemaId
    msg.tag = this.tag
    msg.revocationDetails = this.revocation
    msg = this._addThread(msg)
    return msg
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #write
     */
  async writeMsgPacked (context) {
    return this.getMessageBytes(context, await this.writeMsg(context))
  }

  /**
     * Directs verity-application to write the specified Credential Definition to the Ledger
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async write (context) {
    await this.sendMessage(context, await this.writeMsgPacked(context))
  }
}
module.exports = WriteCredentialDefinition
